Application security part 2 (#337)
3mdeb/news-and-ideas: blog/content/post/2020-10-19-security-part-2.md

---

> title: Application Security as a Every Developer's Responsibility

Reviewer: ```suggestion
title: Application Security as an Every Developer's Responsibility
```

> abstract: "The issue of web application security is a topic that is becoming more and more popular. Currently, not only large organizations and corporations, but also smaller and smaller businesses are forced by progress to transfer some of their activities to the Internet."

Reviewer: ```suggestion
abstract: "The issue of web application security is becoming more and more popular. Currently, not only large organizations and corporations but also smaller businesses are forced by progress to transfer some of their activities to the Internet."
```

cover: /covers/image-file.png
author: piotr.nowoslawski
layout: post
published: true
date: 2020-10-19
archives: "2020"

tags:
  - security
  - application security
categories:
  - Development
  - App Dev
  - Security
---

> The issue of web application security is a topic that is becoming more and more popular. Currently, not only large organizations and corporations, but also smaller and smaller businesses are forced by progress to transfer some of their activities to the Internet. This process has accelerated even more because of the current popularity of all electronic devices and changes in the way people communicate remotely.

Reviewer: jak wyżej

> While large business entities have been operating in the realities of application security for a long time, smaller entities have very little or no knowledge of how to adapt to these new, demanding conditions without exposing themselves to business risk. Unfortunately, the risk of operating in digital markets only increases. Due to increasingly stringent regulations, the loss or disclosure of data has very serious consequences. These can be financial penalties, loss of reputation or customer trust, which ultimately translates into a bad financial result, and in the worst case it will lead to the collapse of the company.

Reviewer: ```suggestion
While large business entities have been operating in the realities of application security for a long time, smaller entities have very little or no knowledge of adapting to these new, demanding conditions without exposing themselves to business risk. Unfortunately, the risk of operating in digital markets only increases. Due to increasingly stringent regulations, the loss or disclosure of data has severe consequences. These can be financial penalties, loss of reputation, or customer trust, finally turning into a bad financial result. In the worst case, it may lead to the collapse of the company.
```

### Why should we care?

Let's focus on the risk in more detail. At the beginning, we need to be aware of what can happen if a data leak occurs in the company.

The first noticeable effect may be financial penalties related to violation of the law or non-compliance with the industry certification. We operate in an environment of data processing regulations such as [GDPR (General Data Protection Regulation)](https://eugdpr.org/) in the European Union, [HIPAA (Health Insurance Portability and Liability Act)](https://www.hhs.gov/hipaa/for-professionals/privacy/laws-regulations/index.html) in the USA and [PCI DSS (data security standard in the payment card industry)](https://www.pcisecuritystandards.org/pci_security/) around the world. On the Internet, we can find a lot of cases where the amount of fines in relation to violations of these regulations is described. Penalties are counted in thousands of euros, which affects the imagination how they can affect the company. It is also worth adding that the imposed penalties are not the end of costs. Incidental costs such as legal services, courts, etc. must be taken into account.

> Let's say we have already paid the penalty costs. Now we have to make up for these losses somehow, but will our application still be able to earn for itself? If the leak was serious and third parties suffered as a result, there will be a loss of reputation, and hence the loss of regular and new customers. You also have to count on the attack of the competition, which will certainly emphasize the favorable situation for them, as our clients will also learn.

Reviewer: ```suggestion
Let's say we have already paid the penalty costs. As a next step, we have to make up for these losses somehow, but our application still earns for itself? If the leak was serious and third parties have suffered, we may loose the reputation along with the regular and new customers. The situation may be enhanced by competitors, to emphasize the favorable situation for them, as our clients will also learn.
```

> Finally, there will be problems such as our ranking in Google, which will have a negative impact on sales very quickly.

Reviewer: ```suggestion
The overall impact on sales will be supported by decreasing Google ranks.
```

> Google is constantly investigating whether we use best security practices such as TLS / SSL, which are used to encrypt traffic and ensure the confidentiality and integrity of data transmission, as well as server and sometimes customer authentication.

Reviewer: ```suggestion
Google is constantly investigating whether we use the best security practices such as TLS / SSL, which are used to encrypt traffic and ensure the confidentiality and integrity of data transmission, server, and sometimes customer authentication.
```

The bottom line is that there is no argument that would challenge making security a priority. Disregarding it does not pay off, and the seemingly saved costs may bring about the opposite result than intended.

### How to deal with security challenges?

The first step is the right approach to the software development process.

There are some rules to follow:

- continuous development of your skills through your own development and training within the organization,
- sharing knowledge about safety in the team,

> - taking responsibility for not releasing an underdeveloped product under time pressure,

Reviewer: Can you simplify the sentence a bit?

- Using Components with known vulnerabilities

Today, even simple websites like personal blogs have many dependencies.
We can all agree that not updating every software on the backend and front end of the website will no doubt introduce serious security threats sooner rather than later.

- Insufficient logging and monitoring

The importance of website security cannot be underestimated. There are ways to monitor your site regularly so that you can take immediate action when something happens. Lack of an efficient login and monitoring process can increase the damage caused by a website hack.

[Official website of the organization](https://owasp.org/www-project-top-ten/)

The list is periodically published and importantly, consists only 10 points. This makes it easy to analyze and understand, making it a good start to taking a serious approach to security in your organization.

OWASP list should be included in every software development project.
It is the best set of security requirements and best practices, and most importantly, it works!
Following and implementing the OWASP recommendations has only positive consequences and teaches you to apply the best available practices.

> 

Reviewer: It is ok to describe OWASP, but we need in this  blogpost some concrete examples that will say: "hey, we are the good guys, here in 3mdeb we do take our responsibility. Look at our contirbution, or project or any_thing_that_will_come_to_your_mind. Seek for an inspiiration in blog, or  here: https://opensource.3mdeb.com/
or here https://3mdeb.com/events/ or elsewhere. Prove, that we, as 3mdeb do take care about.

---
title: Application Security as a Every Developer's Responsibility
abstract: "The issue of web application security is a topic that is becoming more and more popular. Currently, not only large organizations and corporations, but also smaller and smaller businesses are forced by progress to transfer some of their activities to the Internet."
cover: /covers/image-file.png
author: piotr.nowoslawski
layout: post
published: true
date: 2020-10-19
archives: "2020"

tags:
  - security
  - application security
categories:
  - Development
  - App Dev
  - Security
---

> The issue of web application security is a topic that is becoming more and more popular. Currently, not only large organizations and corporations, but also smaller and smaller businesses are forced by progress to transfer some of their activities to the Internet. The process has accelerated even more due to the current global pandemic and changes in the way people communicate.

Reviewer 2: I'm not sure if we should mention the global pandemic here.

Let's say we have already paid the penalty costs. Now we have to make up for these losses somehow, but will our application still be able to earn for itself? If the leak was serious and third parties suffered as a result, there will be a loss of reputation, and hence the loss of regular and new customers. You also have to count on the attack of the competition, which will certainly emphasize the favorable situation for them, as our clients will also learn.

Finally, there will be problems such as our ranking in Google, which will have a negative impact on sales very quickly.
Google is constantly investigating whether we use best security practices such as TLS / SSL, which are used to encrypt traffic and ensure the confidentiality and integrity of data transmission, as well as server and sometimes customer authentication.

The bottom line is that there is no argument that would challenge making security a priority. Disregarding it does not pay off, and the seemingly saved costs may bring about the opposite result than intended.

> 

Reviewer 2: Maybe we should add here some examples of companies which lost money/pr due to security leaks?

- following external guidelines such a OWASP,
- primarily treating safety as priority, not a necessary evil

The rules are obvious, but how many cases do we really stick to them? In how many cases are we not looking for the best solutions or we do not pay enough attention to tests due to the race against time?

We have to answer these questions ourselves. Being aware of your own approach is the beginning of the road to trying to combine speed and security when developing software.

### What's next? Know your enemy.

Fortunately, the problems we face are similar for all developers and there are organizations that work to describe most of them and teach others how to protect themselves from outside threats. Therefore, we do not have to reinvent the wheel.

##### OWASP (Open Web Application Security Project)

is a nonprofit foundation that works to improve the security of software. [OWASP page](https://owasp.org/)

This organization publishes a list of the most common errors in web applications. This list was established in 2017 and is based on real data obtained from companies, organizations and professionals involved in security testing.

##### The current list for 2020:

> - Injection

Reviewer 2: Probably a short description next to these vulnerabilities might be a good idea.

- Broken Authentication
- Sensitive Data Exposure
- XML External Entities (XXE)
- Broken Access control
- Security misconfigurations
- Cross Site Scripting (XSS)
- Insecure Deserialization
- Using Components with known vulnerabilities
- Insufficient logging and monitoring

The list is periodically published and importantly, consists only 10 points. This makes it easy to analyze and understand, making it a good start to taking a serious approach to security in your organization.

> A detailed description of each issue can be found in this [article](https://sucuri.net/guides/owasp-top-10-security-vulnerabilities-2020/) and on the [official website of the organization](https://owasp.org/www-project-top-ten/)

Reviewer 2: I'm not sure if we should redirect traffic to someone's else article. I would stick with the official owasp website instead.
